### PR TITLE
vSomeIP 3.5.1

### PR DESCRIPTION
**Pull request by fcmonteiro**
- Load Policies Lazy Load
- Test - Processing SD messages with unknown type option
- ensure endpoints before deletion
- Improve "end of file" error handling
- Enable debouncing of events & selective events
- Revert "Test - Processing SD messages with unknown type"
- Logs added to points of failure on registration process
- One *.json to ignorem all
- Someip-tp remote address rework
- Fix crash in multicast_receive receive_cb
- Generate network_test configs directly to build
- Fix deadlock if binding of TCP client endpoint fails
- Added missing includes of iomanip to support compilation on Mint
- Cache not yet registered events
- Return true to make sure endpoints are deleted
- Byteorder implementation
- Reorder of prepare_stop method
- Allows applications in the same process using different security configurations
- Fix to not ignore stop offers when sd acceptance is not required
- Release 3.5.0
=====================

- Restructure Network Tests CMakeLists
- policy.cpp unit test
- Remove deprecated usage of byteorder and use bithelper
- unblock endpoint when closing it
- Find_Debounce_Time made configurable
- Check if configuration_ pointer exists before using
- Initialize routing_state_
- Solved data race in configuration_impl class
- utility.cpp unit tests
- unit tests payload_impl
- unit tests serializer
- unit tests deserializer
- Unit Tests for policy_manager_impl.cpp
- Unit Test - Routing Manager set_routing_state
- Move documentation to markdown
- clang-format to verify the code vsomeip-lib
- fix deadlock with event and message debounce feature
- disabled some groups of tests
- Try to force connection reset on suspend
- Call availability handler on request service instance
- Handle endpoint queue size underflow
- Add Valgrind massif tool
- Create new train after scheduling to avoid duplicate messages
- Re-Added offer_tests group in all sanitizers tests
- removed extra DLT logs of the policies print
- COVESA-615: vsomeip.lck file not removed upon application termination
- COVESA-527: Locally switch off -Wstringop-overflow
- Create network regression test for specific issue
- Change IndentPPDirectives rule in clang-format
- applied auto in some identified lines by sonarqube
- Add Valgrind memcheck
- Ensure buffer is valid before de-referencing pointer
- Renaming folder test and fixing typos
- Fix minor formatting issues from some commits
- Adds interger overflow check
- Adds application name on cout logs
- Adding helgrind, to test output
- Boost 1.65 cleanup
- network_test - Offer Stop Offer test
- Support host name (env) for internal TCP communication
- Remove cached configuration after app stops
- Optimize tests/network-tests/CMakeLists.txt
- Check if pointer exists before dereferencing it
- avoid requiring valgrind locally
- Fix Lock-Order-Inversion in policy_manager_impl
- Revert "Fix to not ignore stop offers when sd acceptance is not required"
- Fix timeout on offer_stop_offer test
- Restore config_plugin_impl mutex
- Rework [STOP_]OFFER command handling
- some-tp memory consumption increasing fast
- Remove dlt traces from memory_test
- restart_routing_test enabled
- Fix cyclic events
- Tracing LOI
- improve connection log on error path
- added subscribe_notify groups to non-leak verification
- Relocate hostname config command
- allow subscribeACK if at least one offer was sent by SD
- Improve "end of file" error handling
- Update Clang-Format to Version 18
- Debounce tests fix
- Wireshark dissector for vsomeip protocol
- Remove logging on operation cancel in connect_cbk
- Add additional info on failure to open TCP port
- Implementation of SOMEIPSD_00577
- Refactor how niceness values are applied to threads
- prevent race between event expiration/forwarding
- Fix subscribe_notify_one tests
- Fix missing/blocked subscription requests
- change references to C++14 into C++17
- Explicitly check whether an endpoint is in use
- Enabled all network tests with whitelist
- Fix target client id in local_send
- remove redundants package import definitions
- run unit tests on windows
- Fix debounce network tests
- type upgrade and temporary disable of test for QNX build
- Force abort hanging detached threads
- Application tests fix
- Remove behavior from catch block in ~message
- Stop/Start (network) endpoints on suspend/resume
- Reduce the number of copy operations on event payloads
- Sets linger to 0 in local tcp clients
- Prevent exception re-throw in ~message
- remove linger on local_tcp
- Fix android traces build
- Introduce stateful availability handler
- Update the availability state
- Reintroduces the TIME_WAIT for ltcei
- Faster handlers lookup
- Increase app registration timeout
- Force endpoint restart if re registering
- fix semaphore logs
- fix compile issue with pthreads in android
- Disabling set routing state unit test
- Fixing get_policy_manager error with security disabled

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+# This is the clang-format configuration style to be used by Qt,
+# based on the rules from https://wiki.qt.io/Qt_Coding_Style and
+# https://wiki.qt.io/Coding_Conventions
+
+---
+# Webkit style was loosely based on the Qt style
+BasedOnStyle: WebKit
+
+Standard: c++14
```
Comment (kheaactua) [inline .clang-format:14]: This was changed to c++17 in https://github.com/COVESA/vsomeip/commit/cf497232adf84f55947f7a24e1b64e04b49f1f38

**Comment by kheaactua** [comment]
I love that a clang-format file is being added!

File: .gitignore [hunk 2]
```diff
@@ -1,6 +1,7 @@
 /CMakeFiles
 /build*/*
 /examples/hello_world/build
+/.idea/
```
Comment (kheaactua) [inline .gitignore:4]: Can you also add a few other common IDE/build artifacts?  I have a PR for this too but it's still in a very early draft state.
```
/.cache
/compile_commands.json
.ninja
Makefile
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -362,7 +380,11 @@ endif ()
 target_link_libraries(${VSOMEIP_NAME}-e2e ${VSOMEIP_NAME} ${Boost_LIBRARIES} ${USE_RT} ${DL_LIBRARY} ${SystemD_LIBRARIES})
 
 if(${CMAKE_SYSTEM_NAME} MATCHES "QNX")
+    target_link_directories(${VSOMEIP_NAME} PUBLIC "${QNX_TARGET}/${CPUVARDIR}/io-sock/lib")
```
Comment (kheaactua) [inline CMakeLists.txt:383]: I'm not sure how well this generalizes to different QNX deployments.  I would recommend that QNX linker paths be handled by a toolchain managed by the user (outside of this repo)

File: CMakeLists.txt [hunk 4]
```diff
@@ -622,24 +644,6 @@ else()
...
-    message("source-highlight found")
-    add_custom_command(TARGET doc
-                       POST_BUILD
-                       COMMAND asciidoc
-                               -a version=${VSOMEIP_VERSION}
-                               -b html
-                               -o documentation/vsomeipUserGuide.html
-                               ${PROJECT_SOURCE_DIR}/documentation/vsomeipUserGuide)
-endif()
-
```
Comment (kheaactua) [inline CMakeLists.txt:642]: Why was this removed?  I have some small changes that make it actually work (produce a PDF)

(formatting is different because I ran cmake-format on it)
```
find_program(ASCIIDOC_PATH asciidoc)
find_program(SOURCE_HIGHLIGHT_PATH source-highlight)
if("${ASCIIDOC_PATH}" STREQUAL "ASCIIDOC_PATH-NOTFOUND")
  message(WARNING "asciidoc is not installed. Readme can not be built. This may require a pip3 install asciidoc3")
elseif("${SOURCE_HIGHLIGHT_PATH}" STREQUAL "SOURCE_HIGHLIGHT_PATH-NOTFOUND")
  message(WARNING "source-highlight is not installed. Readme can not be built.")
else()
  message("asciidoc found")
  message("source-highlight found")
  add_custom_command(
    TARGET doc
    POST_BUILD
    COMMAND asciidoc -a version=${VSOMEIP_VERSION} -b html -o documentation/vsomeipUserGuide.html
            ${PROJECT_SOURCE_DIR}/documentation/vsomeipUserGuide
  )
endif()

find_program(A2X_PATH a2x)
if(NOT A2X_PATH)
  message(WARNING "a2x is not installed. Readme can not be built as a pdf.")
else()
  add_custom_command(
    TARGET doc
    POST_BUILD
    COMMAND
      a2x -v --no-xmllint -a version=${VSOMEIP_VERSION} -f pdf -a tabsize=4 --dblatex-opts
      '-P latex.output.revhistory=0 -P imagedata.default.scale=maxwidth=10cm,maxheight=8cm' -D
      ${PROJECT_BINARY_DIR}/documentation ${PROJECT_SOURCE_DIR}/documentation/vsomeipUserGuide
    BYPRODUCTS ${PROJECT_BINARY_DIR}/documentation/vsomeipUserGuide.pdf
  )
endif()
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+Our code style is based on the WebKit style, which is in turn influenced by the Qt style.
+
+Webkit style guidelines can be found here [Webkit.org](https://www.webkit.org/code-style-guidelines/)
+
+#### Standard
+
+```yaml
+Standard: c++14
+```
+The C++ standard to be used, in this case, C++14.
```
Comment (kheaactua) [inline CONTRIBUTING.md:70]: c++17 (see https://github.com/COVESA/vsomeip/commit/cf497232adf84f55947f7a24e1b64e04b49f1f38 )

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+Below is the configuration option used in our Clang Format setup:
+
+#### Clang-Format Customization
+
+```yaml
+BasedOnStyle: WebKit
+Standard: c++14
```
Comment (kheaactua) [inline CONTRIBUTING.md:78]: C++ 17 (see https://github.com/COVESA/vsomeip/commit/cf497232adf84f55947f7a24e1b64e04b49f1f38 )

**Comment by kheaactua** [comment]
These weren't touched in this PR, but Aram's team found several bad reference captures that should probably be addressed in this PR if not a separate PR.

I can't upload a patch file here, so embedding it:
```patch
From 6c77e2c07080e68272f180c5b235b0c4e9f4f5f9 Mon Jul 15 15:42:35 2024 -0400
From: Matthew Russell <matthew.g.russell@gmail.com>
Date: Mon, 15 Jul 2024 15:42:35 -0400
Subject: [PATCH] Fix capturing references to stack variables

---
 implementation/configuration/src/configuration_impl.cpp   | 3 +++
 implementation/endpoints/src/tcp_client_endpoint_impl.cpp | 6 +++---
 2 files changed, 6 insertions(+), 3 deletions(-)

diff --git a/implementation/configuration/src/configuration_impl.cpp b/implementation/configuration/src/configuration_impl.cpp
index 65adb0e2..b693163a 100644
--- a/implementation/configuration/src/configuration_impl.cpp
+++ b/implementation/configuration/src/configuration_impl.cpp
@@ -132,6 +132,7 @@ configuration_impl::configuration_impl(const std::string &_path)
 #endif
       logfile_("/tmp/vsomeip.log"),
       loglevel_(vsomeip_v3::logger::level_e::LL_INFO),
+      is_suppress_events_enabled_(false),
       is_sd_enabled_(VSOMEIP_SD_DEFAULT_ENABLED),
       sd_protocol_(VSOMEIP_SD_DEFAULT_PROTOCOL),
       sd_multicast_(VSOMEIP_SD_DEFAULT_MULTICAST),
@@ -251,6 +252,8 @@ configuration_impl::configuration_impl(const configuration_impl &_other)
     sd_port_ = _other.sd_port_;
     sd_protocol_ = _other.sd_protocol_;
 
+    is_suppress_events_enabled_ = _other.is_suppress_events_enabled_;
+
     sd_initial_delay_min_ = _other.sd_initial_delay_min_;
     sd_initial_delay_max_ = _other.sd_initial_delay_max_;
     sd_repetitions_base_delay_= _other.sd_repetitions_base_delay_;
diff --git a/implementation/endpoints/src/tcp_client_endpoint_impl.cpp b/implementation/endpoints/src/tcp_client_endpoint_impl.cpp
index 2a4f0356..703b057f 100644
--- a/implementation/endpoints/src/tcp_client_endpoint_impl.cpp
+++ b/implementation/endpoints/src/tcp_client_endpoint_impl.cpp
@@ -268,7 +268,7 @@ void tcp_client_endpoint_impl::receive() {
         its_recv_buffer = recv_buffer_;
     }
     auto self = std::dynamic_pointer_cast< tcp_client_endpoint_impl >(shared_from_this());
-    strand_.dispatch([self, &its_recv_buffer](){
+    strand_.dispatch([self, its_recv_buffer](){
         self->receive(its_recv_buffer, 0, 0);
     });
 }
@@ -775,7 +775,7 @@ void tcp_client_endpoint_impl::receive_cbk(
             }
             its_lock.unlock();
             auto self = std::dynamic_pointer_cast< tcp_client_endpoint_impl >(shared_from_this());
-            strand_.dispatch([self, &_recv_buffer, _recv_buffer_size, its_missing_capacity](){
+            strand_.dispatch([self, _recv_buffer, _recv_buffer_size, its_missing_capacity](){
                 self->receive(_recv_buffer, _recv_buffer_size, its_missing_capacity);
             });
         } else {
@@ -947,7 +947,7 @@ void tcp_client_endpoint_impl::send_cbk(boost::system::error_code const &_error,
                 if (its_entry.first) {
                     auto self = std::dynamic_pointer_cast< tcp_client_endpoint_impl >(shared_from_this());
                     strand_.dispatch(
-                        [self, &its_entry]() { self->send_queued(its_entry);}
+                        [self, its_entry]() mutable { self->send_queued(its_entry);}
                     );
                 }
             }
-- 
2.45.2
```

File: CMakeLists.txt (hunk 4 above)
Comment (fcmonteiro) [reply CMakeLists.txt:642]: We removed this because the documentation was moved to markdown

File: .clang-format (hunk 1 above)
Comment (fcmonteiro) [reply .clang-format:14]: I will check with the team

**Comment by fcmonteiro** [comment]
> These weren't touched in this PR, but Aram's team found several bad reference captures that should probably be addressed in this PR if not a separate PR.

To be honest I would prefer to have this as a separate PR.
This one is already quite big.

File: CMakeLists.txt (hunk 4 above)
Comment (kheaactua) [reply CMakeLists.txt:642]: oh, awesome.

File: tools/wireshark_plugin/README.md [hunk 7]
```diff
@@ -0,0 +1,13 @@
+# vsomeip-dissector
+Wireshark dissector for vSomeip internal communication via TCP
+
+## How To Use
+
+1. Place `vsomeip-dissector.lua` file in `~/.config/wireshark/plugins/vsomeip/vsomeip-dissector.lua`
+   (create `plugins` directory if it doesn't exist)
+2. In wireshark go to `Analyze` > `Reload Lua Plugins`
+3. In wireshark go to `Analyze` > `Enable Protocols` and search for `vsomeip` and enable it
```
Comment (kheaactua) [inline tools/wireshark_plugin/README.md:9]: I followed these steps, is there any configuration for it?  I see that `vsomeip` is now a protocol in my display filters, but I don't see it in the protocol list in the list of protocols (along sign SOME/IP and the 1000 other ones.)

Also, when I filter on vsomeip, I get no results.  With SOME/IP I had to configure the port.  When I forcefully tell it to decode as vsomeip (adding `decode_as_entry: tcp.port,30510,SOME/IP,VSOMEIP` in `.config/wireshark/decode_as_entries`) then nothing shows up.

Comment (RuiG19) [reply tools/wireshark_plugin/README.md:9]: Hello @kheaactua 
No extra configurations should be needed.
This plugin will only work for the vsomeip internal communication via tcp.
External someip messages will still be dissected by the default someip protocol on wireshark.

for the internal vsomeip communication via tcp, it should look something like this:
image
image

Comment (kheaactua) [reply tools/wireshark_plugin/README.md:9]: Oh, that explains why I didn't see anything, the pcap file I used only had external TCP (all local was done with UDS.)

This is pretty cool.

Thanks for the info!

File: .clang-format (hunk 1 above)
Comment (fcmonteiro) [reply .clang-format]: done in https://github.com/COVESA/vsomeip/pull/733/commits/d452c10902b95ffac25c651fe0bc8f2a68ef4515

File: CONTRIBUTING.md (hunk 5 above)
Comment (fcmonteiro) [reply CONTRIBUTING.md]: done in https://github.com/COVESA/vsomeip/pull/733/commits/d452c10902b95ffac25c651fe0bc8f2a68ef4515

File: CONTRIBUTING.md (hunk 6 above)
Comment (fcmonteiro) [reply CONTRIBUTING.md]: done in https://github.com/COVESA/vsomeip/pull/733/commits/d452c10902b95ffac25c651fe0bc8f2a68ef4515

File: examples/hello_world/CMakeLists.txt [hunk 8]
```diff
@@ -10,7 +10,7 @@ find_package(Threads REQUIRED)
 
 set(VSOMEIP_NAME "vsomeip3")
 
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17")
```
Comment (kheaactua) [inline examples/hello_world/CMakeLists.txt:13]: This works, but it's a deprecated way of doing it.  It should be:
```cmake
target_compile_features(<target> PRIVATE cxx_std_17)
```

Comment (duartenfonseca) [reply examples/hello_world/CMakeLists.txt:13]: added a fix for this here: https://github.com/COVESA/vsomeip/pull/752

File: implementation/service_discovery/src/service_discovery_impl.cpp [hunk 9]
```diff
@@ -1129,6 +1130,7 @@ service_discovery_impl::on_message(
     if(is_suspended_) {
```
Comment (kheaactua) [inline implementation/service_discovery/src/service_discovery_impl.cpp:1130]: Wouldn't this be a good place to use `std::scoped_lock`?

**Comment by kheaactua** [comment]
hey @fcmonteiro

I get build errors on gcc and clang:

```
implementation/runtime/src/application_impl.cpp:1513:46: error: could not convert ‘vsomeip_v3::configuration::get_policy_manager()’ from ‘shared_ptr<vsomeip_v3::policy_manager_impl>’ to ‘shared_ptr<vsomeip_v3::policy_manager>’
 1513 |     return configuration_->get_policy_manager();
      |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
      |                                              |
      |                                              shared_ptr<vsomeip_v3::policy_manager_impl>
```

File: implementation/runtime/src/application_impl.cpp [hunk 10]
```diff
@@ -1530,7 +1530,11 @@ std::shared_ptr<configuration> application_impl::get_configuration() const {
 }
 
 std::shared_ptr<policy_manager> application_impl::get_policy_manager() const {
+#ifndef VSOMEIP_DISABLE_SECURITY
     return configuration_->get_policy_manager();
+#endif
+    VSOMEIP_ERROR << __func__ << ": Must not be called with security disabled.";
```
Comment (kheaactua) [inline implementation/runtime/src/application_impl.cpp:1536]: Shouldn't this be a `VSOMEIP_FATAL`?

Comment (duartenfonseca) [reply implementation/runtime/src/application_impl.cpp:1536]: i will check this

**Comment by duartenfonseca** [comment]
@kheaactua could you try it now, does it work correctly?

**Comment by kheaactua** [comment]
> @kheaactua could you try it now, does it work correctly?

hey @fcmonteiro, it appears to work, I just built this with gcc and clang with `-DVSOMEIP_DISABLE_SECURITY`

Even the `build_tests` target seemed to build now that I've changed from gtest 1.10 to 1.12

File: implementation/runtime/src/application_impl.cpp (hunk 10 above)
Comment (duartenfonseca) [reply implementation/runtime/src/application_impl.cpp:1536]: @kheaactua I ended up changing this to a warning instead of an error, since trying to get the manager isn't exactly an error, but could lead to one if this case is not properly handled. 
If in some instance you use it, it shouldn't create any big issues

**Comment by kheaactua** [comment]
Any chance https://github.com/COVESA/vsomeip/pull/734 can be integrated (full or in part) into this branch? :)

**Comment by duartenfonseca** [comment]
> Any chance #734 can be integrated (full or in part) into this branch? :)
Sorry @kheaactua it will have to be on the next version. i will make sure it will happen. in any case, the next versions will take less time be merged

**Comment by kheaactua** [comment]
> @kheaactua can you prepare a PR with these changes? [#733 (comment)](https://github.com/COVESA/vsomeip/pull/733#issuecomment-2252743279)

I pushed commit https://github.com/kheaactua/vsomeip-sd/commit/2f5ffe42771d78c0629b58804f8ea61d250c95a1 but I can't seem to target a PR to https://github.com/fcmonteiro/vsomeip/tree/vsomeip_3.5.x

File: implementation/utility/include/qnx_helper.hpp [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#ifdef __QNX__
+#ifndef VSOMEIP_V3_QNX_HELPER_HPP_
+#define VSOMEIP_V3_QNX_HELPER_HPP_
+#include <sys/socket.h>
+#include <netinet/in.h>
+
+ #define SO_BINDTODEVICE 0x0800		/* restrict traffic to an interface */
+ #define IP_PKTINFO		25   /* int; send interface and src addr */
```
[the remark this answers is not in the document]
Comment (duartenfonseca) [reply implementation/utility/include/qnx_helper.hpp:13]: hey @danielfr-haleytek this changes were done because our QNX build was failing, for the reasons you mentioned. The PR you mentioned will be tested on our side, we will give feedback there.
